Add bank customer managing UI tests

Unweldify opened the pull request:
## Реализован автотест панели менеджера банка для управления клиентами.
### В проекте имплементированы:
### Сьют тестов `test_customer`:
- Добавление клиента.
- Удаление клиента из списка.
- Сортировка списка клиентов.

### `Allure` для читабельности и описания шагов тестов.

<hr>

### Эта ветка будет слияться с веткой `main`, где собственно и будет находиться финальный проект автотестов.

Unweldify committed:
Сделан проект
Добавлены docstring
Интегрирован allure

Unweldify committed:
Обновил README

Unweldify committed:
Добавил библиотеки в req.txt

StasovSt commented on Pages/base_page.py:
```diff
@@ -0,0 +1,37 @@
+from selenium.webdriver.remote.webelement import WebElement
+from selenium.webdriver.support import expected_conditions
+from selenium.webdriver.support.wait import WebDriverWait
+from typing import List
```
typing относится к встроенной библиотеке - перенеси его наверх. Прошу пройтись по проекту и поправить импорты в соответствии с pep-8

```
Импорты должны быть сгруппированы в следующем порядке:
- импорты из стандартной библиотеки
- импорты сторонних библиотек
- импорты модулей текущего проекта 
Вставляйте пустую строку между каждой группой импортов
```

Unweldify replied:
Понял, исправил
В следующий раз учту

StasovSt commented on Pages/add_customer_page.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class AddCustomerPage(BasePage):
+    """Класс страницы добавления клиента, наследуется из BasePage."""
```
Описывать в комментарии от кого класс наследуется - излешне

Unweldify replied:
Чутка увлекся, убрал

StasovSt commented on Tests/test_run.py:
```diff
@@ -0,0 +1,65 @@
+
+
+@allure.title('Проверить функциональность добавления клиента')
+def test_1(driver) -> None:
```
Предлагать тестовым функция дать более осмысленные наименования

Unweldify replied:
Обновил наименования

Unweldify committed:
Исправил некоторые недочёты docstring
Дал тест кейсам более осмысленные названия
Импорт библотек по стандарту PEP8

Unweldify committed:
Исправил названия тестов (не собирались pytest)

StasovSt commented on README.md:
```diff
@@ -1,2 +1,40 @@
 # UI-SDET-1
 Практикум SDET simbirsoft: задание UI
```
Добавить инструкцию по локальному запуску автотестов

склонировать репозиторий
установить виртуальное окружение
установить зависимости pip instal...
запустить автотесты ...
для просмотра отчета выполнить команду ....

StasovSt commented on Helpers/generator.py:
```diff
@@ -0,0 +1,25 @@
+from typing import Tuple
+
+
+class Generator:
```
Логика реализованая в данном модуле относится к подготовке тестовых данных - поэтому лучше перенести данный модуль в ./data/client
Generator -> Client
generate_info -> generate_client_data

StasovSt commented on Pages/customers_page.py:
```diff
@@ -0,0 +1,48 @@
+        self.__delete_buttons = (By.XPATH, '//button[@ng-click="deleteCust(cust)"]')
+
+    @allure.step("Нажать на кнопку имён два раза для сортировки по алфавиту.")
+    def sort_alphabetically(self) -> None:
```
Попробуй сделать метод более универсальным: в зависимости от входного аргумента - сортировать по или против алфавита

StasovSt commented on Tests/test_run.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@allure.title('Проверить функциональность добавления клиента.')
+def test_1_adding_customer(driver) -> None:
```
test_adding_customer (без цифр), собственно модуль переименовать ./test_customer

StasovSt replied:
В тестовых функция обычно не анотируют, что она возвращает. Лучше убрать

StasovSt commented on conftest.py:
```diff
@@ -0,0 +1,14 @@
+    options.page_load_strategy = 'eager'
+    options.add_argument('--enable-javascript')
+    _driver = webdriver.Chrome(options=options)
+    _driver.get('https://globalsqa.com/angularJs-protractor/BankingProject/#/manager')
```
вынести url из фикстуры в файл-конфиг, например в .env

StasovSt commented on Tests/test_run.py:
```diff
@@ -0,0 +1,67 @@
+
+@allure.title('Проверить функциональность добавления клиента.')
+def test_1_adding_customer(driver) -> None:
+    manager_page = ManagerPage(driver)
```
чтобы в тесте напрямую не объявлять объекты классов пейджей можно как поступить:

Создаешь conftest в директории ./tests
В нем создаешь фикстуры manager_page, add_customer_page и т.д.
эти фикстуры возвращают объект класса, например:
```
@pytest.fixture()
def manager_page(driver)
   return ManagerPage(driver)
```
Для других вызовов по аналогии.
4. Эти фикстуры уже используешь в тесте и можешь обращаться к ним

StasovSt commented on Tests/test_run.py:
```diff
@@ -0,0 +1,67 @@
+    manager_page = ManagerPage(driver)
+    manager_page.to_add_cust()
+
+    # Добавляем пользователя с данными из генератора
```
Вместо комментариев, добавь конструкцию: 
with allure.step("Добавление пользователя"), with allure.step("Проверка добавления ползователя") и т.д в других тестах

StasovSt commented on Helpers/generator.py:
```diff
@@ -0,0 +1,25 @@
+            first_name += chr(97 + int(post_code[i:i+2]) % 26)
+        # В задании не сказано сгенерировать фамилию, но
+        # для создания клиента она обязятелна, поэтому ставлю свою
+        return (first_name, "Egorov", post_code)
```
лучше не харкодить формирование фамилии, а передавать из теста.

Unweldify committed:
Рефактор кода
Вынес урл ссылку из корневой фикстуры в .env
Перенес вызовы классов из теста в фикстуры
Убраны аннотации типа данных из функций тестов
Убрал хардкод фамилии клиента
Переименовал модуль тестов
Убрал лишнее из названий функций тестов
Переделал комменты в шаги аллюром
Реструктурировал папку модулей хелперов
Обновил README

Unweldify committed:
Merge branch 'main' into UI_pytest

StasovSt commented on Tests/test_customer.py:
```diff
@@ -0,0 +1,69 @@
+        customer = generate_client_data()
+
+    # Я не знаю как 'не хардкодить генерацию фамилии', по заданию не сказано генерить фамилию
+    # но фамилия является обязятельной строкой поэтому я вставлю свою
```
в данном случае, ты передаешь фамилию кастомера из теста, а не жестко зашиваешь ее в функции генерации данных. Такое приемлемо. Можно также использовать библиотеку Faker, чтобы она генерировала фамилию

StasovSt commented on Helpers/Client/generator.py:
```diff
@@ -0,0 +1,24 @@
+from typing import Tuple
+
+
+class Generator:
```
Лучше этот модуль в отдельную директорию ./test_data/client

Unweldify committed:
Перенос генератора данных в отдельную директорию
